File: Services/TodoService.cs
using TodoDemo.Models;

namespace TodoDemo.Services;

public class TodoService
{
    private static readonly List<Todo> _todos = new();
    private static int _nextId = 1;
    private static readonly object _lock = new();

    public List<Todo> GetAll()
    {
        return _todos.OrderByDescending(t => t.CreatedAt).ToList();
    }

    public Todo? GetById(int id)
    {
        return _todos.FirstOrDefault(t => t.Id == id);
    }

    public void Add(Todo todo)
    {
        lock (_lock)
        {
            todo.Id = _nextId++;
            todo.CreatedAt = DateTime.Now;
            _todos.Add(todo);
        }
    }

    public void Update(Todo todo)
    {
        lock (_lock)
        {
            var existing = GetById(todo.Id);
            if (existing != null)
            {
                existing.Title = todo.Title;
                existing.Description = todo.Description;
                existing.IsCompleted = todo.IsCompleted;
            }
        }
    }

    public void Delete(int id)
    {
        lock (_lock)
        {
            var todo = GetById(id);
            if (todo != null)
            {
                _todos.Remove(todo);
            }
        }
    }

    public void ToggleComplete(int id)
    {
        lock (_lock)
        {
            var todo = GetById(id);
            if (todo != null)
            {
                todo.IsCompleted = !todo.IsCompleted;
            }
        }
    }
}

File: java-todo-app/src/main/java/com/tododemo/service/TodoService.java
package com.tododemo.service;

import com.tododemo.model.Todo;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.locks.ReentrantLock;
import java.util.stream.Collectors;

@Service
public class TodoService {
    private final List<Todo> todos = new ArrayList<>();
    private int nextId = 1;
    private final ReentrantLock lock = new ReentrantLock();

    /**
     * Get all todos sorted by creation date (newest first)
     */
    public List<Todo> getAll() {
        return todos.stream()
                .sorted(Comparator.comparing(Todo::getCreatedAt).reversed())
                .collect(Collectors.toList());
    }

    /**
     * Get a todo by its ID
     */
    public Optional<Todo> getById(int id) {
        return todos.stream()
                .filter(t -> t.getId() == id)
                .findFirst();
    }

    /**
     * Add a new todo
     */
    public void add(Todo todo) {
        lock.lock();
        try {
            todo.setId(nextId++);
            todo.setCreatedAt(LocalDateTime.now());
            todos.add(todo);
        } finally {
            lock.unlock();
        }
    }

    /**
     * Update an existing todo
     */
    public void update(Todo todo) {
        lock.lock();
        try {
            Optional<Todo> existing = getById(todo.getId());
            if (existing.isPresent()) {
                Todo existingTodo = existing.get();
                existingTodo.setTitle(todo.getTitle());
                existingTodo.setDescription(todo.getDescription());
                existingTodo.setCompleted(todo.isCompleted());
            }
        } finally {
            lock.unlock();
        }
    }

    /**
     * Delete a todo by its ID
     */
    public void delete(int id) {
        lock.lock();
        try {
            todos.removeIf(t -> t.getId() == id);
        } finally {
            lock.unlock();
        }
    }

    /**
     * Toggle the completion status of a todo
     */
    public void toggleComplete(int id) {
        lock.lock();
        try {
            Optional<Todo> todo = getById(id);
            todo.ifPresent(t -> t.setCompleted(!t.isCompleted()));
        } finally {
            lock.unlock();
        }
    }
}
